src/command/subgraph/publish.rs

```diff
@@ -74,4 +112,186 @@ impl Publish {
+        // if a --routing-url is provided AND the URL is unparsable,
+        // we need to warn and prompt the user, else we can assume a publish
+        if let Some(routing_url) = maybe_invalid_routing_url {
+            match Url::parse(&routing_url) {
+                Ok(parsed_url) => {
+                    tracing::debug!("Parsed URL: {}", parsed_url.to_string());
+                    if !vec!["http", "https"].contains(&parsed_url.scheme()) {
+                        Self::prompt_for_publish(
+                            format!("The `{}` protocol is not supported by router, it expects either `http` or `https`. Continuing the publish will make this subgraph unreachable by your supergraph. Would you still like to publish?", &parsed_url.scheme()).as_str(),
+                            stdin,
+                            stderr,
+                            is_atty,
+                        )?;
+                        Self::issue_warnings_non_tty(is_atty, stderr)?;
+                    } else if let Some(host) = parsed_url.host_str() {
+                        if vec!["localhost", "127.0.0.1"].contains(&host) {
+                            Self::prompt_for_publish(
+                                format!("The host `{}` is not routable via the public internet. Continuing the publish will make this subgraph reachable in local development only. Would you still like to publish?", host).as_str(),
+                                stdin,
+                                stderr,
+                                is_atty,
+                            )?;
+                            Self::issue_warnings_non_tty(is_atty, stderr)?;
+                        }
+                    }
+                }
+                Err(parse_error) => {
+                    tracing::debug!("Parse error: {}", parse_error.to_string());
+                    Self::prompt_for_publish(
+                        format!("`{}` is not a valid routing URL. Continuing the publish will make this subgraph unreachable by your supergraph. Would you still like to publish?", routing_url).as_str(),
+                        stdin,
+                        stderr,
+                        is_atty,
+                    )?;
+                    Self::issue_warnings_non_tty(is_atty, stderr)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn prompt_for_publish(
+        message: &str,
+        reader: &mut impl std::io::Read,
+        writer: &mut impl std::io::Write,
+        is_atty: bool,
+    ) -> RoverResult<Option<bool>> {
+        if !is_atty {
+            Ok(None)
+        } else {
+            write!(writer, "{} [y/N] ", message)?;
+            let mut response = [0];
+            reader.read_exact(&mut response)?;
+            if std::str::from_utf8(&response).unwrap().to_lowercase() == *"y" {
+                Ok(Some(true))
+            } else {
+                Err(anyhow!("Publish cancelled by user").into())
+            }
+        }
+    }
```

Comment:
@EverlastingBugstopper this fn is fairly duplicative of the prompt default no util, but I did want the ability to stub the reader/writer. Happy to port this change to the util and update usages (I think there's only 1).

Reply:
porting would be great! would you mind doing so in a follow-up after we merge?

Reply:
Follow-up sounds great!